File: haskell/glue/src/Glue/Lib/Bool/Maybe.hs
module Glue.Lib.Bool.Maybe where

import Glue.Eval (Eval, apply, eval, throwError)
import Glue.Eval.Exception (wrongNumberOfArguments)
import Glue.IR (IR (..), exists)

maybe_ :: IR Eval
maybe_ = Special maybeImpl

maybeImpl :: [IR Eval] -> Eval (IR Eval)
maybeImpl [f, x] = do
    x' <- eval x
    if exists x'
        then do
            g <- eval f
            apply g [x']
        else pure Void
maybeImpl _ = throwError wrongNumberOfArguments

File: dart/glue/lib/src/lib/bool/maybe.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

final Ir maybe = IrSpecial(maybeImpl);

Eval<Ir> maybeImpl(List<Ir> args) {
  return switch (args) {
    [final f, final x] => eval(x).bind((x_) {
      return isExist(x_)
          ? eval(f).bind((g) => apply(g, [x_]))
          : Eval.pure(IrVoid());
    }),
    _ => throwError(wrongNumberOfArguments()),
  };
}
